cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java

```diff
@@ -143,7 +148,24 @@ protected AbstractContext(Program program, ProgramOptions programOptions,
                                                   program.getApplicationSpecification().getPlugins());
     this.admin = new DefaultAdmin(dsFramework, new NamespaceId(program.getId().getNamespace()), secureStoreManager);
     this.secureStore = secureStore;
-    this.transactional = Transactions.createTransactional(getDatasetCache());
+    this.defaultTxTimeout = determineTransactionTimeout();
+    this.transactional = Transactions.createTransactional(getDatasetCache(), defaultTxTimeout);
+  }
+
+  /**
+   * Be default, this parses runtime argument "system.tx.timeout". Some program types will override this,
+   * for example, in a flowlet, the more specific "flowlet.[name].system.tx.timeout" would prevail.
+   * @return the default transaction timeout, if specified in the runtime arguments, or null otherwise.
+   */
+  protected Integer determineTransactionTimeout() {
+    Integer txTimeout = null;
+    String txTimeoutArg = getRuntimeArguments().get(Constants.Transaction.ARGUMENT_TX_TIMEOUT);
+    try {
+      txTimeout = txTimeoutArg == null ? null : Integer.parseInt(txTimeoutArg);
+    } catch (NumberFormatException e) {
+      LOG.warn("Ignoring value {} for system.data.tx.timeout because it cannot be parsed as an integer.", txTimeoutArg);
+    }
+    return txTimeout;
```

Comment:
Seems like `txTimeout` can be `null`. In that case, the method needs to be annotated with `@Nullable`. Or shall we always default it to the value in cConf?